sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -189,6 +189,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
     // plan may contain nodes that do not report stats. Anything that uses stats must run after
     // this batch.
     Batch("Early Filter and Projection Push-Down", Once, earlyScanPushDownRules: _*) :+
+    // This batch contains rules that should be applied to writes early. For example,
+    // we have to construct a logical write early so that we can inject needed repartition/sort
+    // operators to satisfy data source distribution and ordering requirements.
+    // Expression optimizations must be run before this batch so that we have optimal
+    // expressions when we construct writes. At the same time, rules that dedup repartition and
+    // sort operators must by run afterwards.
+    Batch("Early Writes", Once, earlyWriteRules: _*) :+
```

Comment:
Should we combine `earlyScanPushDownRules` and `earlyWriteRules`? I think they are compatible. We could just come up with a better name for this phase, like `v2SourceRewriteRules`.

Reply:
I think they may be more rules like this in the future, not just writes. I am definitely +1 on making this more flexible. Let me do this in a separate PR.

Reply:
I am resolving this as this was implemented in #30558.